### PR TITLE
Lgz training program

### PR DESCRIPTION
**Pull request by a8520238**
我增加了programs组件

### CODE REVIEW

**Review by a8520238 — COMMENTED** [review]
some lint programs

File: src/app/components/sidebar/sidebar.component.ts [hunk 1]
```diff
@@ -16,6 +16,7 @@ export const ADMIN_ROUTE_ITEMS: RouteInfo[] = [
 export const REGULAR_USER_ROUTE_ITEMS: RouteInfo[] = [
     { path: '/dashboard', title: '首页', icon: 'dashboard', class: '' },
     { path: '/training-record/entry', title: '培训记录填报', icon: 'create', class: '' },
+    { path: '/training-program/programs', title: '培训项目', icon: 'create', class: '' },
```
Comment (Time1ess) [inline src/app/components/sidebar/sidebar.component.ts:20]: `title` should be '培训项目查看'

File: src/app/interfaces/department.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+export interface Department {
+    id: number;
+    create_time: string;
+    up_time: string;
+    name: string;
+    admins: [];
```
Comment (Time1ess) [inline src/app/interfaces/department.ts]: The type of `admins` should be `number[]`

File: src/app/interfaces/department.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export interface Department {
+    id: number;
+    create_time: string;
+    up_time: string;
```
Comment (Time1ess) [inline src/app/interfaces/department.ts]: Wrong name.

File: src/app/interfaces/category.ts [hunk 4]
```diff
@@ -0,0 +1,4 @@
+export interface Category {
```
Comment (Time1ess) [inline src/app/interfaces/category.ts]: Name should be consistent with backend.

File: src/app/interfaces/program.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { Department } from './department';
+import { Category } from './category';
+export interface ProgramDetail {
```
Comment (Time1ess) [inline src/app/interfaces/program.ts]: Wrong name.

File: src/app/interfaces/program.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+    id: number;
+    department_detail: Department;
+    category_detail: Category;
+    form: [];
```
Comment (Time1ess) [inline src/app/interfaces/program.ts]: Type.

File: src/app/modules/training-program/components/programs/programs.component.html [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    <div class="col-12">
+      <div class="card">
+        <div class="card-body">
+          <div class="loading-shade" *ngIf="isLoadingResults">
+            <mat-spinner></mat-spinner>
+          </div>
+          <div class="table-responsive-lg">
+            <table class="table table-hover text-center">
+              <thead class="text-primary d-flex">
+                <th class="col-2 col-sm-1"></th>
```
Comment (Time1ess) [inline src/app/modules/training-program/components/programs/programs.component.html]: The sum of col sizes should no more than 12.

File: src/app/modules/training-program/components/programs/programs.component.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+@Component({
+  selector: 'app-programs',
+  templateUrl: './programs.component.html',
+  styleUrls: ['./programs.component.css']
+})
+export class ProgramsComponent implements OnInit {
+  /** The data to be displayed */
+  programs: ProgramDetail[] = [];
+  /** How to sort the columns */
+  displayedColumns: string[] = ['name', 'department'];
```
Comment (Time1ess) [inline src/app/modules/training-program/components/programs/programs.component.ts]: You are not using `MatTable`, the `displayedColumns` isn't needed.

File: src/app/modules/training-program/services/program.service.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import { ProgramDetail } from 'src/app/interfaces/program';
+import { AUTH_SERVICE, AuthService } from 'src/app/interfaces/auth-service';
+import { PaginatedResponse } from 'src/app/interfaces/paginated-response';
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ProgramService {
+  /** How often should we query latest programforms. */
```
Comment (Time1ess) [inline src/app/modules/training-program/services/program.service.ts]: Delete comment.

File: src/app/modules/training-program/services/program.service.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+  constructor(
+    private readonly http: HttpClient,
+    @Inject(AUTH_SERVICE) private readonly authService: AuthService,
+  ) {
+    this.authService.authenticationSucceed.subscribe(() => {
```
Comment (Time1ess) [inline src/app/modules/training-program/services/program.service.ts]: Why do you need this line?

**Review by Time1ess — CHANGES_REQUESTED** [review]
Revert changes on package-lock.json.

File: src/app/modules/training-program/components/programs/programs.component.html [hunk 11]
```diff
@@ -10,12 +10,12 @@
             <table class="table table-hover text-center">
               <thead class="text-primary d-flex">
                 <th class="col-2 col-sm-1"></th>
```
Comment (Time1ess) [inline src/app/modules/training-program/components/programs/programs.component.html]: The error isn't fixed.

**Review by Time1ess — CHANGES_REQUESTED** [review]

**Comment by a8520238** [comment]
我已经修复了column问题
image
效果如上

**Comment by Time1ess** [comment]
1. Normally, we would make sure that the sum of column sizes are equal to 12, please update.
2. You haven't revert the changes on `package-lock.json`.

**Comment by a8520238** [comment]
现在我回退了版本，并更改了之前的错误
image
页面如上

**Comment by a8520238** [comment]
修改命名

**Comment by a8520238** [comment]
修改类型和排版

**Comment by a8520238** [comment]
类型修改

**Review by dutlzn — DISMISSED** [review]

**Comment by Time1ess** [comment]
@a8520238 The branch is out-dated.

File: src/app/modules/training-program/components/programs/programs.component.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+})
+export class ProgramsComponent implements OnInit {
+  /** The data to be displayed */
+  programs: Program[] = [];
+  /** The total number of notifications. */
+  programsLength = 0;
+  /** Indicate data loading status */
+  isLoadingResults = true;
+
+  pageSize = 5;
```
Comment (Time1ess) [inline src/app/modules/training-program/components/programs/programs.component.ts]: Use global constant instead of 5.

**Review by Time1ess — CHANGES_REQUESTED** [review]

**Review by Time1ess — APPROVED** [review]

**Review by dutlzn — APPROVED** [review]

**Review by Yqzzxlj — APPROVED** [review]

**Pull request merged**